src/linters.ts

```diff
@@ -0,0 +1,164 @@
+) => void
+
+type LinterDefinition = {
+  linterFunc: LinterFunc
+}
+
+const linters: Record<string, LinterDefinition> = {
+  muon: {
+    linterFunc: lintDocumentMuon
+  }
+}
+
+async function fileIsExecutable(filePath: string): Promise<boolean> {
+  return new Promise<boolean>(resolve => {
+    fs.access(filePath, fs.constants.X_OK, err => resolve(err ? false : true));
+  });
+}
+
+async function findExecutable(name: string): Promise<string | undefined> {
+  for (const dir of (process.env['PATH'] || '').split(path.delimiter)) {
+    const file = path.join(dir, name);
+
+    if (await fileIsExecutable(file)) {
+      return file;
+    }
+  }
+
+  return undefined;
+}
+
+async function execLinter(
+  command: string,
+  args: string[],
+  options: cp.ExecOptions = {},
+  stdin: string
+): Promise<{ stdout: string; stderr: string, error?: cp.ExecFileException }> {
+  return new Promise<{ stdout: string; stderr: string, error?: cp.ExecFileException }>(resolve => {
+    let p = cp.execFile(command, args, options, (error, stdout, stderr) => {
+      const err = error ? error : undefined;
+
+      resolve({ stdout, stderr, error: err });
+    });
+
+    p.stdin?.write(stdin);
+    p.stdin?.end();
+  });
+}
+
+async function lintDocumentMuon(
+  muon_path: string,
+  root: string,
+  diagnosticCollection: vscode.DiagnosticCollection,
+  document: vscode.TextDocument
+) {
+  diagnosticCollection.clear();
+
+  const { error, stdout, stderr } = await execLinter(
+    muon_path,
+    ["analyze", "-l", "-O", document.uri.fsPath],
+    { cwd: root },
+    document.getText()
+  )
+
+  let diagnostics: vscode.Diagnostic[] = [];
+  stderr.split("\n").forEach(line => {
+    const parts = line.split(":");
+    if (parts.length < 4) {
+      return;
+    }
+
+    const file = parts[0];
+    const line_no = Number(parts[1]);
+    const col = Number(parts[2]);
+    const fullmsg = parts.slice(3).join(":").trim();
+
+    if (file != document.uri.fsPath) {
+      return;
+    }
+
+    const range = new vscode.Range(line_no - 1, col, line_no - 1, col);
+
+    let severity: vscode.DiagnosticSeverity = vscode.DiagnosticSeverity.Error;
+    if (fullmsg.startsWith("warn")) {
+      severity = vscode.DiagnosticSeverity.Warning;
+    }
+
+    const msg = fullmsg.slice(fullmsg.indexOf(" ") + 1);
+
+    const diagnostic = new vscode.Diagnostic(range, msg, severity);
+    diagnostics.push(diagnostic);
+  });
+
+  diagnosticCollection.set(document.uri, diagnostics);
+}
+
+async function reloadLinters(sourceRoot: string, context: vscode.ExtensionContext): Promise<vscode.Disposable[]> {
+  let disposables = [];
+
+  for (const [name, props] of Object.entries(linters)) {
+    const config: LinterConfiguration = extensionConfiguration("linters")[name];
+
+    if (!config.enabled) {
+      continue;
+    }
+
+    let linter_path = config.path;
+    if (!linter_path) {
+      linter_path = await findExecutable(name);
+    }
+
+    if (!linter_path) {
+      continue;
```

Comment:
If it's enabled and cannot be found we should likely create an error message (notification/popup) about that.